src/coreclr/jit/gentree.cpp

```diff
@@ -13589,6 +13589,30 @@ GenTree* Compiler::gtFoldExprSpecial(GenTree* tree)
         return icon;
     };
 
+    auto NewZeroExtendNode = [&](var_types type, GenTree* op1, var_types castToType) -> GenTree* {
+        assert(varTypeIsIntegral(type));
+        assert(!varTypeIsSmall(type));
+        assert(!varTypeIsUnsigned(type));
+        assert(varTypeIsUnsigned(castToType));
+
+        GenTreeCast* cast = gtNewCastNode(TYP_INT, op1, false, castToType);
+        if (fgGlobalMorph)
+        {
+            fgMorphTreeDone(cast);
+        }
+
+        if (varTypeIsLong(type))
+        {
+            cast = gtNewCastNode(TYP_LONG, cast, true, TYP_ULONG);
```

Comment:
this doesn't have to be ULONG, right? cc @SingleAccretion

Reply:
Yes we actually prefer it to be `TYP_LONG`.

Also the `varTypeIsLong` checks can be just `type == TYP_LONG` as well.

Reply:
I see, so it doesn't quite matter if the `castToType` is `TYP_LONG` or `TYP_ULONG`.

The `type == TYP_LONG` should work as well since `type` can never be unsigned.